File: 21-Binary-Search-Tree-Part-II/python/01_floor_and_ceil_in_a_bst.py
from typing import List, Optional

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def floorCeilOfBST(self, root: Optional[TreeNode], key: int) -> List[int]:
        def find_floor(node: Optional[TreeNode], k: int) -> int:
            floor = -1
            while node:
                if node.val == k:
                    return node.val
                elif node.val < k:
                    floor = node.val
                    node = node.right
                else:
                    node = node.left
            return floor

        def find_ceil(node: Optional[TreeNode], k: int) -> int:
            ceil = -1
            while node:
                if node.val == k:
                    return node.val
                elif node.val > k:
                    ceil = node.val
                    node = node.left
                else:
                    node = node.right
            return ceil

        return [find_floor(root, key), find_ceil(root, key)]

File: 21-Binary-Search-Tree-Part-II/java/01-FloorAndCeilInABst.java
import java.util.*;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> floorCeilOfBST(TreeNode root, int key) {
        int floor = findFloor(root, key);
        int ceil = findCeil(root, key);
        List<Integer> result = new ArrayList<>();
        result.add(floor);
        result.add(ceil);
        return result;
    }

    private int findFloor(TreeNode root, int key) {
        int floor = -1;
        while (root != null) {
            if (root.val == key) {
                return root.val;
            } else if (root.val < key) {
                floor = root.val;
                root = root.right;
            } else {
                root = root.left;
            }
        }
        return floor;
    }

    private int findCeil(TreeNode root, int key) {
        int ceil = -1;
        while (root != null) {
            if (root.val == key) {
                return root.val;
            } else if (root.val > key) {
                ceil = root.val;
                root = root.left;
            } else {
                root = root.right;
            }
        }
        return ceil;
    }
}
